[AArch64] Lower mathlib call ldexp into fscale when sve is enabled

huhu233 opened the pull request:
The function of 'fscale' is equivalent to mathlib call ldexp, but has better performance. This patch lowers ldexp into fscale when sve is enabled.

nikic commented:
Why does this need special handling in CGP instead of being a normal custom lowering for FLDEXP?

paulwalker-arm commented:
I agree and custom lowering also gives a straight forward way to support vectors types as well.

huhu233 commented:
> I agree and custom lowering also gives a straight forward way to support vectors types as well.

You are right, I'll update the patch as soon, thanks!

huhu233 commented:
> Why does this need special handling in CGP instead of being a normal custom lowering for FLDEXP?

@nikic , thanks for your suggestion, I'll update the patch as soon!

huhu233 commented:
* Rebase and update the patch

davemgreen reviewed (commented):
Hi. Can you split this into two separated patches - one for AArch64 and another for X86? I think they should be logically separable.

arsenm commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -57389,3 +57396,40 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
+  Exp = DAG.getNode(ISD::SINT_TO_FP, DL, XScalarTy, Exp);
+  SDValue VX =
+      DAG.getNode(ISD::INSERT_VECTOR_ELT, DL, XVT, DAG.getUNDEF(XVT), X, Zero);
+  SDValue VExp = DAG.getNode(ISD::INSERT_VECTOR_ELT, DL, ExpVT,
```
Is this lowering the scalar case as a vector operation? Should this move to the vector legalizer?

huhu233 replied:
Hi, @arsenm, here are some of may consideration:

* Is this lowering the scalar case as a vector operation?
Actually, `ldexp@PLT` (ISD::`FLDEXP`) is a scalar call, but `llvm.aarch64.sve.fscale.*` or `llvm.x86.mask.scalef.*` only has vector versions. I hope to replace call with more efficient target supported instructions, so there are some `INSERT_VECTOR_ELT` and `EXTRACT_VECTOR_ELT`. 

* "Should this move to the vector legalizer?" 
`FLDEXP` will occur in many scalar cases, so I think it is not necessary to move to the vector legalizer.

huhu233 commented on llvm/lib/Target/X86/X86ISelLowering.cpp:
```diff
@@ -57389,3 +57396,40 @@ Align X86TargetLowering::getPrefLoopAlignment(MachineLoop *ML) const {
+  SDLoc DL(Op);
+  EVT XVT, ExpVT;
+  SDValue IID;
+  switch (Op.getValueSizeInBits()) {
```
Thanks, I'll fix it.

huhu233 commented on llvm/test/CodeGen/AArch64/ldexp.ll:
```diff
@@ -0,0 +1,55 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py UTC_ARGS: --version 3
+; RUN: llc -O3 -mtriple=aarch64 -mattr=+sve < %s -o - | FileCheck %s
+
+define dso_local nofpclass(nan inf) double @testExp(double noundef nofpclass(nan inf) %val, i32 noundef %a) {
```
OK, thanks for your suggestions!

huhu233 commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -26360,3 +26367,39 @@ bool AArch64TargetLowering::preferScalarizeSplat(SDNode *N) const {
+    XVT = MVT::nxv4f32;
+    ExpVT = MVT::nxv4i32;
+    break;
+  case 64:
```
You are right, but it seems `f16` for `ldexp` is not fully supported? I got some compile failures as shown, 
https://godbolt.org/

huhu233 commented:
* Rebase the branch
* Split the original patch into 2 seperate commits
* Clean up related testcases
* Support `f16` scenario.

huhu233 commented:
ping

davemgreen commented:
Sorry when I said patch I should have said pull request. Can you split the two patches into separate prs, so they can be reviewed separately. I think what you have for AArch64 looks OK to me

huhu233 commented:
> Sorry when I said patch I should have said pull request. Can you split the two patches into separate prs, so they can be reviewed separately. I think what you have for AArch64 looks OK to me

Hi, @davemgreen , thanks for your reply. Do you mean split these two commits into two Pull Request? If so, I'll remove the commit about X86 in this Patch, and create another Pull Request for it.

davemgreen commented:
Yeah that sounds good to me, thanks.

huhu233 commented:
* rebase
* Split X86 code into another Pull Request.

davemgreen reviewed (commented):
Thanks. I checked the edge cases and this looks good to me, if you can address the last round of comments.

huhu233 committed:
[AArch64] Lower mathlib call ldexp into fscale when sve is enabled

The function of 'fscale' is equivalent to mathlib call ldexp, but has
better performance. This patch lowers ldexp into fscale when sve is
enabled.

huhu233 commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -26414,3 +26422,46 @@ bool AArch64TargetLowering::preferScalarizeSplat(SDNode *N) const {
+  default:
+    return SDValue();
+  case MVT::f16:
+    X = DAG.getNode(ISD::FP_EXTEND, DL, MVT::f32, X);
```
Done, thank very much!

huhu233 commented on llvm/test/CodeGen/AArch64/ldexp.ll:
```diff
@@ -0,0 +1,66 @@
+
+declare half @llvm.ldexp.f16.i32(half, i32) #1
+
+attributes #1 = { mustprogress nofree nosync nounwind willreturn memory(none) }
```
Done, only keep the necessary attribute.

huhu233 commented on llvm/lib/Target/AArch64/AArch64ISelLowering.cpp:
```diff
@@ -26414,3 +26422,46 @@ bool AArch64TargetLowering::preferScalarizeSplat(SDNode *N) const {
   return true;
 }
+
+SDValue AArch64TargetLowering::LowerFLDEXP(SDValue Op,
```
Done, thanks!

davemgreen reviewed (approved):
Thanks, this LGTM.

Do you have commit access, or should I hit commit? If so are you happy for this to go in?

huhu233 commented:
> Thanks, this LGTM.
> 
> Do you have commit access, or should I hit commit? If so are you happy for this to go in?

Hi, @davemgreen, I have asked someone to hit commit, thanks for your reviews!
